packages/driver/src/httpScram.ts

```diff
@@ -41,27 +66,42 @@ export async function HTTPSCRAMAuth(
       Authorization: `SCRAM-SHA-256 data=${utf8ToB64(clientFirst)}`,
     },
   });
-  if (serverFirstRes.status === 403) {
-    // tslint:disable-next-line:no-console
-    console.log(serverFirstRes);
-    throw new Error(`Server doesn't support HTTP SCRAM authentication`);
+
+  // The first request must have status 401 Unauthorized and provide a
+  // WWW-Authenticate header with a SCRAM-SHA-256 challenge.
+  // See: https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L153-L157
+  const authenticateHeader = serverFirstRes.headers.get("WWW-Authenticate");
+  if (serverFirstRes.status !== 401 || !authenticateHeader) {
+    const body = await serverFirstRes.text();
+    throw new ProtocolError(`authentication failed: ${body}`);
   }
-  const firstAttrs = parseHeaders(serverFirstRes.headers, "WWW-Authenticate");
-  if (firstAttrs.size === 0) {
-    throw new Error("Invalid credentials");
+
+  // WWW-Authenticate can contain multiple comma-separated authentication
+  // schemes (each with own comma-separated parameter pairs), but we only support
+  // one SCRAM-SHA-256 challenge, e.g., `SCRAM-SHA-256 sid=..., data=...`.
+  if (!authenticateHeader.startsWith("SCRAM-SHA-256")) {
+    throw new ProtocolError(
+      `unsupported authentication scheme: ${authenticateHeader}`
+    );
   }
-  if (!firstAttrs.has("sid") || !firstAttrs.has("data")) {
+
+  // The server may respond with a 401 Unauthorized and `WWW-Authenticate: SCRAM-SHA-256` with
+  // no parameters if authentication fails, e.g., due to an incorrect username.
+  // See: https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L112-L120
```

Comment:
Love these helpful comments! I'll probably update them to the relevant parts of [RFC 7804](https://datatracker.ietf.org/doc/html/rfc7804#section-5) since they're less likely to line-number sensitive and we'd like to think of the client and server being decoupled at least the protocol level here.

Reply:
I was thinking of doing the same! Unfortunately, I couldn't find an example of this specific behaviour in the RFC — did I miss it?

Reply:
Yeah, I think this just falls under the server returning a response that doesn't conform to the standard and thus should be rejected rather than something explicitly spelled out in the spec itself.

I think this is fine in lieu of having a more specified thing to point at!